access log: add support for command formatter extensions

rgs1 opened the pull request:
Fixes #13411.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

rgs1 commented:
The windows failure seems transient:

```
2020/12/23 18:07:34 Downloading https://releases.bazel.build/3.6.0/release/bazel-3.6.0-windows-x86_64.exe...
2020/12/23 18:07:45 could not download Bazel: HTTP GET https://releases.bazel.build/3.6.0/release/bazel-3.6.0-windows-x86_64.exe failed: Get "
Deleting service account key file...

```

rgs1 commented:
/retest

qiwzhang commented on test/common/formatter/substitution_formatter_test.cc:
```diff
@@ -2354,6 +2354,57 @@ TEST(SubstitutionFormatterTest, ParserSuccesses) {
+  StreamInfo::MockStreamInfo stream_info;
+  std::string body;
+
+  bool command_found;
```
please initialize it as false.

rgs1 replied:
@qiwzhang can we enable clang-tidy to detect these things?

qiwzhang commented on source/common/formatter/substitution_formatter.cc:
```diff
@@ -282,116 +288,141 @@ void SubstitutionFormatParser::parseCommand(const std::string& token, const size
+      // Try extensions.
+      bool added = false;
+      for (const auto& cmd : commands) {
+        auto formatter = cmd->parse(token);
```
Do we need to pass `pos` to the cmd->parse?  Otherwise  how will extension handle extra fields, such as `path`, `field_length`?

rgs1 replied:
Ah yes, good point

qiwzhang commented on source/common/formatter/substitution_formatter.cc:
```diff
@@ -282,116 +288,141 @@ void SubstitutionFormatParser::parseCommand(const std::string& token, const size
+      formatters.push_back(std::move(formatter));
+    } else {
+      // Try extensions.
+      bool added = false;
```
Could you clarify when to use format extensions?  
1) just the format change, such as unquoted_dynamic_metadata,   this is the same data as dynamic_metadata, but the format is different.
2) new data,  for example, https://github.com/envoyproxy/envoy/issues/14230.  should we use extensions

mattklein123 replied:
+1 for more docs per my above comment

mattklein123 commented on api/envoy/config/core/v3/substitution_format_string.proto:
```diff
@@ -16,9 +17,18 @@ option (udpa.annotations.file_status).package_version_status = ACTIVE;
 
+// Specifies the configuration for a formatter extension.
+message Formatter {
+  string name = 1 [(validate.rules).string = {min_len: 1}];
```
For new extensions does the name field serve any purpose? And would we ever have any other types of config? I'm wondering if typed_config can be used in the repeated below and we can just lookup the factory based on type_url? I think we do that for new extensions or maybe I keep getting this confused?

rgs1 replied:
So I mostly followed https://github.com/envoyproxy/envoy/pull/13676, which is pretty recent. It's mostly the same:

```
  repeated FatalAction fatal_actions = 28;

...
  message FatalAction {
    // Extension specific configuration for the action. It's expected to conform
    // to the ``Envoy::Server::Configuration::FatalAction`` interface.
    core.v3.TypedExtensionConfig config = 1;
  }
```

Let me check what other extension points do.

I do think we could end up with different types of configs... cc: @dio who recently reviewed another formatter that should be an extension.

mattklein123 replied:
`TypedExtensionConfig` is what I was thinking of. I think you can just do repeated `TypedExtensionConfig` below and that should be fine? We should be looking up by type_url also so see other examples for that (maybe that is already what is being done). Thank you!

/wait

rgs1 replied:
Ok np, #10908 did it that way. I'll follow that.

mattklein123 commented on include/envoy/formatter/substitution_formatter.h:
```diff
@@ -78,5 +79,34 @@ class FormatterProvider {
+public:
+  virtual ~CommandParser() = default;
+
+  virtual FormatterProviderPtr parse(const std::string& token) const PURE;
```
Please add comments for the class above and comments for this function. I had to read the code below to figure out how a format extension works. Please be detailed.

mattklein123 commented on source/common/formatter/substitution_formatter.h:
```diff
@@ -60,6 +61,13 @@ class SubstitutionFormatParser {
                            std::vector<std::string>& sub_items, absl::optional<size_t>& max_length);
 
+  /**
+   *  Handle builtin commands.
```
Can you add doc comments? This code is kind of confusing from a quick read.

mattklein123 reviewed (changes requested):
Thanks nice to see this finally implemented. A few comments to get started.

/wait

rgs1 commented:
@mattklein123 I've addressed your comments -- PTAL. 

One remaining question: should we add some extensions to `/source/extensions/formatters`? We could move some of the built-in extensions there or maybe add something like the dynamic_metadata_unquoted extension from #12991?

This would ensure that we are using/exercising the new extension point... Thoughts?

mattklein123 commented on docs/root/version_history/current.rst:
```diff
@@ -64,6 +64,7 @@ Removed Config or Runtime
 
 New Features
 ------------
+* access log: added an extension point for custom formatters (command operators).
```
Maybe ref link to the new repeated field?

mattklein123 commented on include/envoy/formatter/substitution_formatter.h:
```diff
@@ -78,5 +79,47 @@ class FormatterProvider {
+   * @return FormattterProviderPtr substitution provider for the parsed command
+   */
+  virtual FormatterProviderPtr parse(const std::string& token, size_t pos,
+                                     int command_end_position) const PURE;
```
nit: is there any reason to use int here vs. size_t?

rgs1 replied:
not really, fixed

mattklein123 commented on include/envoy/formatter/substitution_formatter.h:
```diff
@@ -78,5 +79,47 @@ class FormatterProvider {
+  /**
+   * Return a FormatterProviderPtr if this command is parsed from the token.
+   * @param token the token to parse
+   * @param pos current position in the entire format string
+   * @param command_end_position position at the end of the command token
```
I'm not very familiar with the formatting code, but it's pretty confusing. Can you provide a small example of what token/pos/command_end_position might actually be?

rgs1 replied:
added something

mattklein123 commented on source/common/formatter/substitution_formatter.cc:
```diff
@@ -282,116 +288,143 @@ void SubstitutionFormatParser::parseCommand(const std::string& token, const size
+      }
+
+      if (!added) {
         formatters.emplace_back(FormatterProviderPtr{new StreamInfoFormatter(token)});
```
This will fail at this point if the token is unknown, right?

rgs1 replied:
it will fail inside StreamInfoFormatter()'s constructor if it's unknown, yes

mattklein123 reviewed (changes requested):
Thanks this LGTM overall. Is it possible to add an integration test that uses a fake extensio? IIRC there are access log helpers so I don't think this will be very difficult?

/wait

kyessenov commented on include/envoy/formatter/substitution_formatter.h:
```diff
@@ -78,5 +79,47 @@ class FormatterProvider {
+   * @return CommandParserPtr the CommandParser which will be used in
+   * SubstitutionFormatParser::parse() when evaluating an access log format string.
+   */
+  virtual CommandParserPtr createCommandParserFromProto(const Protobuf::Message& config) PURE;
```
I think you want a message validation context here as well.

rgs1 replied:
You mean within `createCommandParserFromProto()` something like `ProtobufMessage::getStrictValidationVisitor()` should be used or should it be a param for  `createCommandParserFromProto()` itself?

Looking at this other extension point (https://github.com/envoyproxy/envoy/pull/10908), the validator isn't part of creator method that receives a config...

kyessenov replied:
Yeah, I was looking into this recently [here](https://github.com/envoyproxy/envoy/blob/master/source/extensions/rate_limit_descriptors/expr/config.cc#L64), and it seems that validator generally comes from the factory context. A strict validator might be OK for now, not sure what additional benefit the context validator provides.

mattklein123 reviewed (changes requested):
Awesome, thanks for the testing. Check CI?

/wait

rgs1 commented:
> Awesome, thanks for the testing. Check CI?
> 
> /wait

ah yes, hold on for review tho i still need to address a few points.

rgs1 commented:
@mattklein123 assuming CI goes through (I'll ensure it does), this should address all your comments. Thanks!

Other than that, waiting on @kyessenov for one pending item.

rgs1 commented on test/common/formatter/command_extension.cc:
```diff
@@ -0,0 +1,45 @@
+}
+
+std::string TestCommandFactory::name() const { return "envoy.formatter.TestFormatter"; }
+
```
@mattklein123 shall I add this file to this allow list:

https://github.com/envoyproxy/envoy/blob/master/tools/code_format/check_format.py#L66

so that I can call `REGISTER_FACTORY()` from here instead of relying on the InjectFactory helper?

mattklein123 replied:
I think using the helper is better where you need it?

mattklein123 reviewed (changes requested):
Thanks LGTM. Can you merge main to fix the release notes after the release?

/wait

mattklein123 reviewed (dismissed):
Thanks!

mattklein123 commented:
Sorry needs another main merge.

/wait
